sdk/java-sdk/src/main/java/kalix/javasdk/StatusCode.java

```diff
@@ -0,0 +1,41 @@
+/*
+ * Copyright 2021 Lightbend Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package kalix.javasdk;
+
+/**
+ * Interface used to represent a status code, typically used when replying with an error. **NOT**
+ * for user extension.
+ */
+public interface StatusCode {
+  /**
+   * The supported error codes that can be used when replying from the Kalix user function.
+   *
+   * <p>Note: the error code will be transcoded to a respective HTTP or gRPC code depending on the
+   * type of service being exposed.
+   */
+  enum ErrorCode implements StatusCode {
+    BAD_REQUEST,
+    UNAUTHORIZED,
+    FORBIDDEN,
+    NOT_FOUND,
+    CONFLICT,
+    TOO_MANY_REQUESTS,
+    INTERNAL_SERVER_ERROR,
+    SERVICE_UNAVAILABLE,
+    GATEWAY_TIMEOUT
+  }
+}
```

Comment:
What about this...

```java
public interface StatusCode {

  interface ErrorCode extends StatusCode {}
  /**
   * The supported error codes that can be used when replying from the Kalix user function.
   *
   * <p>Note: the error code will be transcoded to a gRPC code depending on the
   * type of service being exposed.
   */
  enum HttpErrorCode implements ErrorCode {
    BAD_REQUEST,
    UNAUTHORIZED,
    FORBIDDEN,
    NOT_FOUND,
    CONFLICT,
    TOO_MANY_REQUESTS,
    INTERNAL_SERVER_ERROR,
    SERVICE_UNAVAILABLE,
    GATEWAY_TIMEOUT
  }
}
```

Eventually, in the future we can introduce:

```java
enum GrpcErrorCode implements ErrorCode {
  ... 
}
```

I would keep the `error()` method accepting `ErrorCode` in case we want to have `GrpcErrorCode`.

Reply:
There is a catch on my suggestion above. 

The javasdk is compiling with jdk11 and we can't use sealed interfaces.

The current impl, prevents new error codes because it's an enum. While my suggestion would make `ErrorCode' and open interface.

Reply:
> There is a catch on my suggestion above.
> 
> The javasdk is compiling with jdk11 and we can't use sealed interfaces.
> 
> The current impl, prevents new error codes because it's an enum. While my suggestion would make `ErrorCode' and open interface.

Yup, been there. :D 

That was exactly the rational. Since we don't have a sealed interface I decided to use make it closed with the Enum.

That said, I think given the current strategy, it might make sense to follow your suggestion and have it explicitely as `HttpErrorCode`. Will update.

Reply:
Oh, there was another reason for the current approach.. my thinking was also: if we change our minds and want to be able to have agnostic internal error codes after all, with the current approach that would still be possible. We would just need to rename the param on `error()` to drop the `http` but it would not need any breaking change for the API, it would still be source-compatible. If we tie the type to Http then I think we have a stronger commitment to have this dual approach supporting HTTP and gRPC.

Reply:
Ok, I see. Let's keep it as is then. 

There is a merge conflict now. I will fix it myself and merge this PR.